File: orchalanguage/src/main/java/orcha/lang/compiler/OrchaCompiler.java
package orcha.lang.compiler;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.boot.CommandLineRunner;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.WebApplicationType;
import org.springframework.context.annotation.ComponentScan;
import org.springframework.context.annotation.Configuration;

import java.util.List;

public class OrchaCompiler {

    private static Logger log = LoggerFactory.getLogger(OrchaCompiler.class);

    @Autowired
    Preprocessing preprocessing;

    @Autowired
    @Qualifier("lexicalAnalysisForOrchaCompiler")
    LexicalAnalysis lexicalAnalysis;

    @Autowired
    SyntaxAnalysis syntaxAnalysis;

    @Autowired
    SemanticAnalysis semanticAnalysis;

    @Autowired
    Postprocessing postprocessing;

    @Autowired
    OutputGeneration outputGeneration;

    public void compile(String orchaFileName) throws OrchaCompilationException {

        List<String> linesOfCode = preprocessing.process(orchaFileName);

        OrchaProgram orchaProgram = lexicalAnalysis.analysis(linesOfCode);

        orchaProgram  = syntaxAnalysis.analysis(orchaProgram);

        orchaProgram  = postprocessing.process(orchaProgram);

        Object output  = outputGeneration.generation(orchaProgram);

    }

}

File: orchalang/src/main/kotlin/orcha/lang/compiler/OrchaCompiler.kt
package orcha.lang.compiler

import org.slf4j.LoggerFactory

class OrchaCompiler(private val preprocessing: Preprocessing, private val lexicalAnalysis: LexicalAnalysis, private val syntaxAnalysis: SyntaxAnalysis, private val semanticAnalysis: SemanticAnalysis, private val postprocessing: Postprocessing, private val linkEditor: LinkEditor, private val outputGeneration: OutputGeneration) {

    @Throws(OrchaCompilationException::class)
    fun compile(orchaFileName: String) {

        val linesOfCode = preprocessing!!.process(orchaFileName)

        var orchaProgram = lexicalAnalysis!!.analysis(linesOfCode)

        orchaProgram = syntaxAnalysis!!.analysis(orchaProgram)

        orchaProgram = semanticAnalysis!!.analysis(orchaProgram)

        orchaProgram = postprocessing!!.process(orchaProgram)

        orchaProgram = linkEditor!!.link(orchaProgram)

        val generatedCode = outputGeneration!!.generation(orchaProgram)


    }

    companion object {
        private val log = LoggerFactory.getLogger(OrchaCompiler::class.java)
    }

}
